altair/vegalite/v5/api.py

```diff
@@ -0,0 +1,2790 @@
+    if context is not None and data_transformers.consolidate_datasets:
+        data = _consolidate_data(data, context)
+
+    # if data is still not a recognized type, then return
+    if not isinstance(data, (dict, core.Data)):
+        warnings.warn("data of type {} not recognized".format(type(data)))
+
+    return data
+
+
+# ------------------------------------------------------------------------
+# Aliases & specializations
+Bin = core.BinParams
+
+
+@utils.use_signature(core.LookupData)
+class LookupData(core.LookupData):
+    def to_dict(self, *args, **kwargs):
+        """Convert the chart to a dictionary suitable for JSON export."""
+        copy = self.copy(deep=False)
+        copy.data = _prepare_data(copy.data, kwargs.get("context"))
+        return super(LookupData, copy).to_dict(*args, **kwargs)
+
+
+@utils.use_signature(core.FacetMapping)
+class FacetMapping(core.FacetMapping):
+    _class_is_valid_at_instantiation = False
+
+    def to_dict(self, *args, **kwargs):
+        copy = self.copy(deep=False)
+        context = kwargs.get("context", {})
+        data = context.get("data", None)
+        if isinstance(self.row, str):
+            copy.row = core.FacetFieldDef(**utils.parse_shorthand(self.row, data))
+        if isinstance(self.column, str):
+            copy.column = core.FacetFieldDef(**utils.parse_shorthand(self.column, data))
+        return super(FacetMapping, copy).to_dict(*args, **kwargs)
+
+
+# ------------------------------------------------------------------------
+# Encoding will contain channel objects that aren't valid at instantiation
+core.FacetedEncoding._class_is_valid_at_instantiation = False
+
+# ------------------------------------------------------------------------
+# These are parameters that are valid at the top level, but are not valid
+# for specs that are within a composite chart
+# (layer, hconcat, vconcat, facet, repeat)
+TOPLEVEL_ONLY_KEYS = {"background", "config", "autosize", "padding", "$schema"}
+
+
+def _get_channels_mapping():
+    mapping = {}
+    for attr in dir(channels):
+        cls = getattr(channels, attr)
+        if isinstance(cls, type) and issubclass(cls, core.SchemaBase):
+            mapping[cls] = attr.replace("Value", "").lower()
+    return mapping
+
+
+# -------------------------------------------------------------------------
+# Tools for working with parameters
+class Parameter(expr.core.OperatorMixin, object):
+    """A Parameter object"""
+
+    _counter = 0
+
+    @classmethod
+    def _get_name(cls):
+        cls._counter += 1
+        return "parameter{:03d}".format(cls._counter)
+
+    def __init__(self, name):
+        if name is None:
+            name = self._get_name()
+        self.name = name
+
+    def to_dict(self):
+        if self.param_type == "variable":
+            if self.param.expr is Undefined:
+                return {"expr": self.name}
+            else:
+                return {"expr": repr(self.param.expr)}
```

Comment:
I've the feeling the behavior seen is because of this line: https://github.com/altair-viz/altair/pull/2528/files#diff-bbe4e187b18d242a366c820d023afd89041759cb96e4ec66c3f34559a72c2f9dR177.